File: Programmers/[60059] 자물쇠와 열쇠/SUbbb/60059.java
import java.util.*;

class Solution {
    static int M;
    static int N;
    static int[][] bigMap;
    
    public boolean solution(int[][] key, int[][] lock) {  
        M = key.length;
        N = lock.length;
        
        for (int r = 0; r < M - 1 + N; r++) {
            for (int c = 0; c < M - 1 + N; c++) {
                for (int rotate = 0; rotate < 4; rotate++) {
                    // 자물쇠를 중앙에 두고, 키를 돌려가면서 확인할 큰 배열 선언
                    bigMap = new int[N + (M - 1) * 2][N + (M - 1) * 2];

                    // 자물쇠를 중앙에 배치
                    for (int i = 0; i < N; i++)
                        System.arraycopy(lock[i], 0, bigMap[M - 1 + i], M - 1, N);

                    // bigMap에 키를 더해 자물쇠에 맞는지 확인할 것
                    match(key, r, c);
                    if (canUnlock()) return true;
                }
            }
        }
        
        return false;
    }
    
    // bigMap이 1로만 이루어지는지 확인
    static private boolean canUnlock() {
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                if (bigMap[M - 1 + i][M - 1 + j] != 1) return false;
        return true;
    }

    // 키와 자물쇠를 더하는 함수
    static private void match(int[][] key, int r, int c) {
        key = rotatingKey(key);

        for (int i = 0; i < M; i++)
            for (int j = 0; j < M; j++)
                bigMap[i + r][j + c] += key[i][j];
    }
    
    // 키를 시계방향으로 90도 돌리는 함수
    static private int[][] rotatingKey(int[][] key) {
        for (int i = 0; i < M / 2; i++) {
            int[] tmp = Arrays.copyOf(key[M - i - 1], M);
            key[M - i - 1] = key[i];
            key[i] = tmp;
        }

        for (int i = 0; i < M; i++) {
            for (int j = i; j < M; j++) {
                int temp = key[i][j];
                key[i][j] = key[j][i];
                key[j][i] = temp;
            }
        }

        return key;
    }
}

File: Programmers/[60059] 자물쇠와 열쇠/JFe/60059.py
def solution(key, lock):
    m, n = len(key), len(lock)
    for angle in range(4):
        # 처음 제외하고, 시계 방향 회전
        if angle != 0:
            temp = [[0 for _ in range(m)]for _ in range(m)]
            for i in range(m):
                for j in range(m):
                    temp[j][m - 1 - i] = key[i][j]
            key = temp
        # key 옮기면서 확인
        for i in range(m + n):
            for j in range(m + n):
                clear = True
                # 큰 배열 grid 만들어서 key 옮기기
                grid = [[-1 for _ in range(n + m * 2)] for _ in range(n + m * 2)]
                for r in range(m):
                    for c in range(m):
                        grid[i+r][j+c] = key[r][c]
                # grid 상 key와 lock 비교
                for r in range(n):
                    for c in range(n):
                        if grid[m-1+r][m-1+c] == -1 and lock[r][c] == 0:    # key가 벗어났는데 그 부분이 자물쇠 홈인 경우
                            clear = False
                        elif grid[m-1+r][m-1+c] != -1 and not grid[m-1+r][m-1+c]^lock[r][c]:   # key와 lock의 값이 같은 경우 (맞지 않는 경우)
                            clear = False
                if clear: 
                    return True
    return False